pkg/controllers/worker/worker_available_controller.go

```diff
@@ -0,0 +1,108 @@
+package worker
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/controllers/common"
+)
+
+var knownConditionNames = sets.NewString(
+	"WorkerNodesAvailable",
+)
+
+// workersAvailableController validates that router certs match the ingress domain
+type workersAvailableController struct {
+	operatorClient v1helpers.OperatorClient
+	nodeLister     corev1listers.NodeLister
+}
+
+func NewWorkersAvailableController(
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder,
+	nodeInformer corev1informers.NodeInformer,
+) factory.Controller {
+	controller := &workersAvailableController{
+		operatorClient: operatorClient,
+		nodeLister:     nodeInformer.Lister(),
+	}
+
+	return factory.New().
+		WithInformers(
+			operatorClient.Informer(),
+			nodeInformer.Informer(),
+		).
+		WithSync(controller.sync).
+		ResyncEvery(30*time.Second).
+		ToController("WorkersAvailableController", eventRecorder)
+}
+
+func countReadyNodes(nodes []*corev1.Node) int {
+	readyNodes := 0
+	for _, n := range nodes {
+		for _, c := range n.Status.Conditions {
+			if c.Type == "Ready" && c.Status == "True" {
+				readyNodes++
+			}
+		}
+	}
+	return readyNodes
+}
+
+func (c *workersAvailableController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	foundConditions := []operatorv1.OperatorCondition{}
+
+	workers, err := c.nodeLister.List(labels.SelectorFromSet(labels.Set{"node-role.kubernetes.io/worker": ""}))
+	if err != nil {
+		return err
+	}
+
+	// this is best-effort, workers can be tainted and not schedulable, which will result in auth operator to fail
+	// as the router need schedulable worker nodes.
+	readyWorkers := countReadyNodes(workers)
+
+	// we don't have any worker nodes schedulable, but we can run clusters that have master nodes schedulable, so we need to check that
+	// before going available==false
+	var masters []*corev1.Node
+	if readyWorkers == 0 {
```

Comment:
can't customers have both? schedulable masters and nodes?

Reply:
good question, I think we can check all nodes then and find if we have something router pods can schedule on.